### PR TITLE
fix auto scale batch size not working with precision=16

### LINKED ISSUE #2655: auto_lr_finder crashes when using 16-bit precision with pytorch-nightly and torchvision-nightly
I heard that the nightly version of pytorch has native support for 16-bit training and wanted to give it a try since I'm trying to train some recent models on a GTX 1080. FYI, I'm using `pytorch-lightning=0.85.0`.

I've installed the following version of the two libraries:
* torch: https://download.pytorch.org/whl/nightly/cu102/torch-1.7.0.dev20200720-cp37-cp37m-linux_x86_64.whl
* torch-vision: https://download.pytorch.org/whl/nightly/cu102/torchvision-0.8.0.dev20200720-cp37-cp37m-linux_x86_64.whl

I've also setup the `Trainer` as follows:

```python
    trainer = Trainer(
        gpus=1,
        max_epochs=hparams.epochs,
        auto_lr_find=True,
        progress_bar_refresh_rate=0,
        accumulate_grad_batches=10,
        # overfit_batches=5,
        amp_level="O2",
        precision=16,
        logger=logger,
        checkpoint_callback=checkpoint_callback,
    )
```

I'm training a resnext101_32x8d_wsl model using the weights provided by Facebook in `pytorch-hub`.

```
Running command:
        python pipe/train_cnn.py
/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/utilities/distributed.py:25: UserWarning: Checkpoint directory /home/gianluca/git/kaggle/siim-isic-melanoma-classification/models exists and is not empty with save_top_k != 0.All files in this directory will be deleted when
a checkpoint is saved!
  warnings.warn(*args, **kwargs)
Using cache found in /home/gianluca/.cache/torch/hub/facebookresearch_WSL-Images_master
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
CUDA_VISIBLE_DEVICES: [0]
Using native 16bit precision.
Traceback (most recent call last):
  File "pipe/train_cnn.py", line 237, in <module>
    main(create_submission=True)
  File "pipe/train_cnn.py", line 48, in main
    preds, weight_fpath = train(fold_number=fold_number, folds=folds)
  File "pipe/train_cnn.py", line 120, in train
    trainer.fit(model)
  File "/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 956, in fit
    self._run_lr_finder_internally(model)
  File "/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/trainer/lr_finder.py", line 58, in _run_lr_finder_internally
    lr_finder = self.lr_find(model)
  File "/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/trainer/lr_finder.py", line 180, in lr_find
    self.save_checkpoint(str(save_path))
  File "/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/trainer/training_io.py", line 268, in save_checkpoint
    checkpoint = self.dump_checkpoint(weights_only)
  File "/home/gianluca/git/kaggle/siim-isic-melanoma-classification/.venv/lib/python3.7/site-packages/pytorch_lightning/trainer/training_io.py", line 362, in dump_checkpoint
    checkpoint['native_amp_scaling_state'] = self.scaler.state_dict()
AttributeError: 'NoneType' object has no attribute 'state_dict'
ERROR: failed to reproduce 'train_cnn.dvc': stage: 'train_cnn.dvc' cmd 'python pipe/train_cnn.py' failed
``` 

 - PyTorch Version (e.g., 1.0): torch-1.7.0.dev20200720
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): poetry
 - Build command you used (if compiling from source): 
 - Python version: 3.7.0
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 1 x GTX 1080
 - Any other relevant information:

### Additional context

Since `torch^1.6.0` has native support to 16-bit training, I did not install NVidia APEX. The whole reason of using a nightly version of pytorch was to avoid to install APEX since I wasn't able to figure out how to install it with `poetry`.

### LINKED ISSUE #3025: Auto-scaling batch-size not compatible with half precision training
Using `precision=16` and `auto_scale_batch_size=True` yields `'NoneType' object has no attribute 'state_dict'` error

### Expected behavior

Largest batch size should be found when using 16 bit precision

### Environment
- PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8
 - CUDA/cuDNN version: 10/7
 - GPU models and configuration: 2080ti

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #3025
Fixes #2655 (we may need an extra test)

Made sure test is failing on master

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: tests/trainer/test_trainer_tricks.py [hunk 1]
```diff
@@ -257,3 +258,19 @@ def test_error_on_dataloader_passed_to_fit(tmpdir):
...
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=1,
+        auto_scale_batch_size=True,
+        gpus=1,
+        precision=16
+    )
+    trainer.fit(model)
+    assert trainer.amp_backend == AMPType.NATIVE
+    assert trainer.scaler is not None
```
Comment (SkafteNicki) [inline tests/trainer/test_trainer_tricks.py:276]: Could we add an assert to check that the batch size have been altered?
Add `before_batch_size = model.batch_size` before training
Add `after_batch_size = model.batch_size` after training
Add `assert before_batch_size != after_batch_size` in the end

Comment (awaelchli) [reply tests/trainer/test_trainer_tricks.py:278]: yep thanks. see new commit

Comment (SkafteNicki) [reply tests/trainer/test_trainer_tricks.py:278]: Nice :+1:

**Review by rohitgr7 — APPROVED** [review]

**Pull request merged**